myia/frontends/pytorch.py

```diff
@@ -172,61 +161,38 @@ def new_module(*args):
                 setattr(v, k, a)
         return v
 
-    return AbstractModule(v.__class__, attrs, {'__call__': fwd_fn,
-                          '__sub__': mod_sub}, constructor=new_module)
+    return AbstractModule(v.__class__, attrs, constructor=new_module)
 
 
 @to_abstract.register  # noqa: F811
 def _to_abstract(self, v: torch.Tensor, **kwargs):
-    return AbstractPyTorchTensor(
+    return AbstractArray(
         AbstractScalar({
             VALUE: ANYTHING,
             TYPE: pytorch_dtype_to_type(v.dtype),
         }),
-        {SHAPE: tuple(v.shape)},
+        {SHAPE: tuple(v.shape), TYPE: PyTorchTensor},
     )
 
 
 @to_abstract.register  # noqa: F811
 def _to_abstract(self, v: torch.nn.Parameter, **kwargs):
-    return AbstractPyTorchTensor(
+    return AbstractArray(
         AbstractScalar({
             VALUE: ANYTHING,
             TYPE: pytorch_dtype_to_type(v.dtype),
         }),
-        {SHAPE: tuple(v.shape)},
+        {SHAPE: tuple(v.shape), TYPE: PyTorchTensor},
     )
 
-##############################################################################
-
 
-@convert_arg.register
-def _convert_arg(self, arg, orig_t: AbstractPyTorchTensor):
-    et = orig_t.element
-    assert isinstance(et, AbstractScalar)
-    et = et.values[TYPE]
-    assert issubclass(et, Number)
+@convert_arg_array.register
+def _convert_arg_array(arg, t: PyTorchTensor, et, orig_t):
```

Comment:
Since you changed the location of the type checks in pipeline/steps.py, they should be present here too.

Alternatively this could only check that it is a pytorch tensor and convert it to numpy, leaving the type/shape checks as common code.

Reply:
Yes, good idea.